src/simple_key_provider.rs

```diff
@@ -140,16 +151,79 @@ impl<R: SafeRng> KeyProvider for SimpleKeyProvider<R> {
     }
 }
 
+#[async_trait]
+impl<R: SafeRng> KeyProvider<U32> for SimpleKeyProvider<U32, R> {
```

Comment:
Comments as above. There's lots of duplication here. Mainly I'm concerned about it because it makes it harder to keep code specifics in sync. Any bugs would need to be fixed in 2 places etc.

Reply:
@ragibkl +1 for not having to have a duplicate implementation for each key size.